src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs

```diff
@@ -192,13 +192,14 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
                     {
                         var entry = item.Entry;
                         var matcher = item.TemplateMatcher;
-                        if (!matcher.TryMatch(context.HttpContext.Request.Path, context.RouteData.Values))
-                        {
-                            continue;
-                        }
 
                         try
                         {
+                            if (!matcher.TryMatch(context.HttpContext.Request.Path, context.RouteData.Values))
```

Comment:
I don't know this code at all, but even though this specific code change looks correct to me (it's now properly implemented as `try { do; } finally { cleanup; }`), why does `matcher.TryMatch()` throw the offending exception to begin with? Is that also a bug that should be fixed?

I'm worried that this change has the correct behavior, but is perhaps masking some other bug.